// angular-file/esm2020/file-upload/ngfSelect.directive.mjs
import { Directive, Input } from "@angular/core";
import { ngf } from "./ngf.directive";
import * as i0 from "@angular/core";
export class ngfSelect extends ngf {
    constructor() {
        super(...arguments);
        this.selectable = true;
    }
}
ngfSelect.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ngfSelect, deps: null, target: i0.ɵɵFactoryTarget.Directive });
ngfSelect.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.3.0", type: ngfSelect, selector: "[ngfSelect]", inputs: { selectable: "selectable" }, exportAs: ["ngfSelect"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ngfSelect, decorators: [{
            type: Directive,
            args: [{
                    selector: "[ngfSelect]",
                    exportAs: "ngfSelect"
                }]
        }], propDecorators: { selectable: [{
                type: Input
            }] } });
